src/refinement/triquadrisection.jl

```diff
@@ -0,0 +1,60 @@
+# ------------------------------------------------------------------
+# Licensed under the MIT License. See LICENSE in the project root.
+# ------------------------------------------------------------------
+
+"""
+    TriQuadrisection()
+
+Refinement of a mesh by preliminarly triangulating it if needed and
+then subdividing each triangle into four triangles.
+
+## References
+
+* Charles Loop. 1987. [Smooth subdivision surfaces based on 
+  triangles](https://charlesloop.com/thesis.pdf). 
+  Master's thesis, University of Utah.
+"""
+struct TriQuadrisection <: RefinementMethod end
+
+function refine(mesh, ::TriQuadrisection)
+  tmesh = eltype(mesh) <: Triangle ? mesh : simplexify(mesh)
+
+  # retrieve vertices
+  points = vertices(tmesh)
+  npoints = length(points)
+
+  # convert topology to half-edge structure
+  t = convert(HalfEdgeTopology, topology(tmesh))
+  ntriangles = nelements(t)
+
+  # add midpoints of edges
+  midpoints = Dict{Tuple{Int,Int},Int}()
+  ∂₁₀ = Boundary{1,0}(t)
+  for eind in 1:nfacets(t)
+    i, j = sort(∂₁₀(eind))
+    midpoint = Segment(points[i], points[j])(0.5)
+    push!(points, midpoint)
+    npoints = npoints + 1
+    midpoints[(i, j)] = npoints
+  end
```

Comment:
I think we can avoid the construction of this dictionary here. We can simply rely on the insertion order. Instead of looping over facets, try to loop over elements (i.e. triangles) and for each of these elements, you insert 4 new points to the end of the vector of points. In a second loop you can then lookup the points in the vector with the correct offsets, and that should be even faster and more memory efficient than then dictionary solution.

Reply:
I'm not convinced. Some faces share a common edge. If I follow the algorithm you outline, if I correctly understand, I will compute multiple times the midpoints of the edges. No?

Reply:
You're right. I had the other refinement method in mind where new points lie inside the elements.